httpcache.go

```diff
@@ -63,32 +76,63 @@ func (Cache) CaddyModule() caddy.ModuleInfo {
 
 // Provision provisions c.
 func (c *Cache) Provision(ctx caddy.Context) error {
-	// TODO: use UsagePool so that cache survives config reloads - TODO: a single cache for whole process?
+	c.logger = ctx.Logger(c)
+
 	maxSize := c.MaxSize
 	if maxSize == 0 {
 		const maxMB = 512
 		maxSize = int64(maxMB << 20)
 	}
 
-	ac, err := autocache.New(&autocache.Options{})
+	if c.Olric.Env == "" {
+		c.Olric.Env = "local"
+	}
+
+	started, cancel := context.WithCancel(context.Background())
```

Comment:
Any reason we aren't using `ctx` here? Using `ctx` will ensure that it will inherit the cancellation from the config, so like if the server is shutting down or the config is changing.

Reply:
I kept this part as-is to be honest. `ctx` looks better suited indeed!

Reply:
Actually this isn't really related here. This context is just used to block while Olric is starting.

Reply:
But what if the server is closed or reloaded while olric is starting? It should still be cancelable right?

Reply:
I handled this in the commit I just pushed.

Reply:
Just so I'm clear, where is it handled? I'm not seeing ctx used...

Reply:
Since the last commit we don't allow to reload Olric, and Olric handles SIGTERM and SIGKILL by itself, so I don't think there is something special to do here.

Reply:
I see, but... what if Caddy is shut down via an API request? Or something other than a signal?

Reply:
Good point, I'm not sure of how to handle that. I'll take a look.

Reply:
Is there a way to detect a shutdown (and not just a reload)?

Reply:
I think you want `caddy.CleanerUpper` https://caddyserver.com/docs/extending-caddy#module-lifecycle

Reply:
@dunglas To clarify, the answer to the question "how to tell the difference between shutdown and reload?" is: there is no difference. If you have global state you need to clean up, the UsagePool essentially does reference counting and will deallocate a value if it reaches 0 references. (As the link Francis posted describes, a reload starts the new module before stopping the old one, so for a short time they overlap.) We also don't guarantee that modules' lifetimes is equivalent to a config lifetime; i.e. a module start/stop != config reload. For example, consider an HTTP handler that might invoke an embedded scripting language (such as Starlark -- which we experimented with early on), the HTTP request will load some modules according to the script, then unload them when the request finishes: a very short lifetime compared to the overall web server config!

Anyway, the right way to think about this problem is to always honor the context lifetime. If possible, signals shouldn't be handled by Olric: insteaad, it should expose methods to call when we have determined we are done with it. (Maybe it does already? I dunno. @buraksezer)

Reply:
So I should re-introduce this code? https://github.com/caddyserver/cache-handler/pull/5/commits/dfc54d14d893312674a9f457c5d8f56997cfd3fd#diff-11c97206e17bbda59c9250af3f2585b8R129

Isn't it better to never shutdown Olric until Caddy itself is shutdown? The cache config is immutable anyway, so it's not useful to stop and start it again during a reload (and it maybe creates problems).

Reply:
@mholt @dunglas Olric doesn't handle any signal in embedded mode. So this shouldn't be a problem for cache-handler module. Signal handling is a part of `olricd`: https://github.com/buraksezer/olric/blob/master/cmd/olricd/server/server.go#L187

> insteaad, it should expose methods to call when we have determined we are done with it.

You just need to call `Shutdown` method if you are done with an Olric instance. It will close all the servers, goroutines and free the allocated space.

> Isn't it better to never shutdown Olric until Caddy itself is shutdown?

Yes, it is. We should never shutdown Olric instance until Caddy itself is shutdown.